### PR TITLE
Revamp docker build process

### PR DESCRIPTION
**Pull request by rbren**
A few changes in here:
* Switch to matrix build for docker images, so we can build them concurrently
* Switch from Makefiles to simple configuration files
* Build on every push
  * branch names will become a tag
    * e.g. `ghcr.io/opendevin/opendevin:main` or `ghcr.io/opendevin/opendevin:rb-docker-tags`
  * git tags will also become docker tags, with semver
    * e.g. `ghcr.io/opendevin/opendevin:v1.2.3` and `ghcr.io/opendevin/opendevin:v1`
  * I believe this will only work for people who can push branches to this repo--maybe we can do this for external PRs too, to make it easier for reviewers to test

Eventually I'd love to automatically add a comment to the PR description that says something like:
> To test this change, run
```bash
docker run ghcr.io/opendevin/opendevin:rb-docker-tags -p 3000:3000
```

### CODE REVIEW

**Comment by foragerr** [comment]
Wouldn't you need a 

```
    permissions:
      packages: write
```

**Comment by foragerr** [comment]
> Build on every push

How much storage does the repo/org get for free?
The current opendevin image size is nearly 3G

**Comment by rbren** [comment]
AFAICT we get unlimited storage 😯 

> GitHub Packages usage is free for public packages. For private packages, each account on GitHub.com receives a certain amount of free storage and data transfer

https://docs.github.com/en/billing/managing-billing-for-github-packages/about-billing-for-github-packages

**Comment by rbren** [comment]
Multi-stage dropped the size down to 1.8G--thanks for the tip @foragerr!

**Comment by rbren** [comment]
OK, this is ready for review! 

Public package is here: https://github.com/opendevin/OpenDevin/pkgs/container/opendevin

**Comment by foragerr** [comment]
Awesome!

File: .github/workflows/ghcr.yml [hunk 1]
```diff
@@ -1,8 +1,8 @@
-name: Build and publish multi-arch container images
+name: Publish Docker Image
 
 on:
   push:
-    branches: [ main ]
+    branches: [ '**' ]
```
Comment (foragerr) [inline .github/workflows/ghcr.yml:5]: At 15+ mins per build, I'm not a fan of "build every commit"
It doesn't block PR merges right?

**Review by foragerr — APPROVED** [review]

File: .github/workflows/ghcr.yml (hunk 1 above)
Comment (rbren) [reply .github/workflows/ghcr.yml:5]: Yeah agree. I'm hopeful some kind of caching will speed this up, but that's a future issue.

It won't block PRs for now.

**Pull request merged**